File: DIA/Aula 04/Arvore de Busca/arvore_de_busca.c
#include <stdio.h>

long long int tree_size(int size, long long* array) {
    if (size == 0) return 1;
    
    else if (array[size-1] != 0) return array[size-1];
    
    else {
        long long int count = 0;

        for (int k=1; k <= size/2; k++) count += 2 * (tree_size(k-1,array) * tree_size(size-k,array));

        count += (size % 2) ? (tree_size(size/2,array) * tree_size(size/2,array)) : 0;

        array[size-1] = count;

        return count;
    }
}

int main() {
    static long long possibilities[35] = {1,2,5};
    
    int t, n;
    scanf("%d",&t);

    for (int i=0; i<t; i++){
        scanf("%d",&n);
        printf("%lld\n",tree_size(n,possibilities));
    }

    return 0;
}

File: DIA/Aula 04/Arvore de Busca/arvore_de_busca.py
def tree_size(size,array):
    if (array[size-1] != 0):
        return array[size-1]
    
    else:
        count = 0
        
        for j in range(size//2):
            count += 2 * (tree_size(j-1,array) * tree_size(size-j,array))

        if (size%2 != 0):
            count += tree_size(size//2,array) ** 2

        return count
    

t = int(input())
vector = [0]*35
vector[0:2] = [1,2,5]

for i in range(t):
    n = int(input())
    print(tree_size(n,vector))
